crates/catalog/src/mutator.rs

```diff
@@ -30,12 +30,52 @@ impl CatalogMutator {
         self.client.as_ref()
     }
 
+    /// Commit the catalog state.
+    /// This persists the state to the metastore.
+    pub async fn commit_state(
+        &self,
+        catalog_version: u64,
+        state: CatalogState,
+    ) -> Result<Arc<CatalogState>> {
+        let client = match &self.client {
+            Some(client) => client,
+            None => return Err(CatalogError::new("metastore client not configured")),
+        };
+
+        let state = match client.commit_state(catalog_version, state).await {
+            Ok(state) => state,
+            Err(CatalogError {
+                msg,
+                strategy: Some(ResolveErrorStrategy::FetchCatalogAndRetry),
+            }) => {
+                // Go ahead and refetch the catalog and retry the mutation.
+                //
+                // Note that this relies on metastore _always_ being stricter
+                // when validating mutations. What this means is that retrying
+                // here should be semantically equivalent to manually refreshing
+                // the catalog and rerunning and replanning the query.
+                debug!(error_message = msg, "retrying mutations");
+
+                client.refresh_cached_state().await?;
+                let state = client.get_cached_state().await?;
```

Comment:
does the `refresh_cached_state` call also apply whatever updates we're trying to commit here? It definitely reads as if we reset the state and then try and commit the state to what it already is. Either this is a bug or one of these methods could be more clearly named.

Reply:
This is an important question to answer. `refresh_cached_state` is responsible for refreshing the cached catalog state on this node by getting the latest version from metastore. `get_cached_state` just gets a reference to it.

Can metastore return uncommitted states? (directly impacting what `refresh_cached_state` will do)

Reply:
added some comments, and cleaned this up